### ISSUE TITLE
Consider our use of `float` and `Decimal`

### ISSUE THREAD

**Issue by jalextowle** [delvtech/elf-simulations#112]
We use `Decimal` only within the confines of the pricing model calculations and use `float` everywhere else. It may be preferable to use `Decimal` throughout the codebase and have a rounding convention of always using a given precision (say 18 decimals of precision) outside of intermediate computations. This would clean up the pricing model implementation by eliminating extraneous casting that is somewhat error-prone.

**Comment by dpaiton** [comment]
We have a few pricing model functions that require high-precision variables to avoid underflow when differences result in values very close to zero. There are two changes I'd like to see:

Propagate changes through the whole repo, so that we can set the precision in config & it holds that precision everywhere (with appropriate rounding on output displays). This will make the code cleaner and also help when we want to test lower precision to compare against solidity contracts.

We solved this using [Decimal](https://docs.python.org/3/library/decimal.html), although the Decimal type does not play well with numpy. Numpy is our primary compute library, so we want to play well with it. We should redo it using [high-precision types from Numpy](https://numpy.org/doc/stable/user/basics.types.html#extended-precision).

**Comment by dpaiton** [comment]
Closed #146 because it is a duplicate of this issue. It has additional details that should be considered when addressing this issue, though.

**Comment by wakamex** [comment]
This issue popped up in the close long and short tests. With high liquidity (500 million) and low trades sizes ($10) we saw cases where the net effect of the round trip trades was zero, meaning that we saw no slippage across these two trades. That shouldn't happen, and it doesn't happen in solidity.

This seems like it may be an issue going forward. But before doing more work, we should investigate the severity. Some tips from @dpaiton:
- is this an issue only for trades on the same block (as is the case with the close long and short tests) or do they also accumulate over time
- how much is the slippage off compared to solidity (cross-codebase testing should identify this)

I put together a quick test of precision that fails only when no slippage is detected, and you can see it fails depending on the precision used by Decimal
![image](https://user-images.githubusercontent.com/16990562/225365765-0c2f15fc-2518-4b87-bc73-37aba481448c.png)

test is here: https://github.com/element-fi/elf-simulations/blob/4ff8083685818f8720fa862c14c10a47bc1e3405/tests/test_precision.py

**Comment by wakamex** [comment]
In the close long test, I got these values:

```
solidity:
base_proceeds: 476190479202948871
    base_paid: 476190482837189261

python:
base_proceeds=0.4761904761960197
    base_paid=0.4761904761960197

diff:
-6.31E-09
-1.39E-08
```

**Comment by wakamex** [comment]
Taking a closer look at the precision tests I created yesterday, slippage seems to go to 0 as precision increases. This makes no sense to me. As well, the slippage on open_long flips sign, which tells me I'm probably measuring it wrong.
![image](https://user-images.githubusercontent.com/16990562/225423531-de5196c8-3a52-49cd-bab0-bf4bb33316d3.png)

sheet with all the results: https://docs.google.com/spreadsheets/d/1MsVhSJ2rI-BasJeiylw7XK5fEe44e2C0HPdGOtOyptM/edit?usp=sharing

tests: https://github.com/element-fi/elf-simulations/pull/305

**Comment by dpaiton** [comment]
Here is an example of how we might use a decorator to enable mix-type support for individual functions. The problem is we are casting back to floats after func calls so we may not actually be able to guarantee a certain precision level:
```
def is_normal(value: float) -> bool:
    """
    Return True if the argument is a _normal_ finite number with
    an adjusted exponent greater than or equal to Emin. Return False
    if the argument is subnormal, infinite or a NaN.
    Note, the term normal is used here in a different sense than typical
    `normalize` methods, which are used to create canonical values.
    """
    return math.isfinite(value) and abs(value) >= sys.float_info.min


def to_fixed_point(float_var: float, decimal_places: int = 18) -> int:
    """Convert floating point argument to fixed point with desired number of decimals"""
    fixed_point = int(float_var * 10**decimal_places)
    # TODO: can I use sys.maxsize?? That's maximum `int`
    # max_value = (10**decimal_places + 1) - 1
    # if fixed_point > sys.maxsize:
    #    raise OverflowError("Fixed point value is too large")
    return fixed_point


def to_floating_point(fixed_var: int, decimal_places: int = 18) -> float:
    """Convert fixed point argument to floating point with specified number of decimals"""
    return float(fixed_var / 10**decimal_places)


T = TypeVar("T", int, float)


def mixed_precision_support(decimal_places: int = 18):
    def decorator(func):
        use_fixed_point = os.environ.get("USE_FIXED_POINT", False)
        if use_fixed_point:

            def fixed_func(*args: Union[float, int]) -> FixedPoint:
                result_fp = func(*[FixedPoint(arg, decimal_places) for arg in args])
                return result_fp

            def wrapper(*args: T) -> T:
                # if inputs were float then do the operations with
                # fixed point and then convert back to float
                result = fixed_func(*args)
                if isinstance(args[0], int):  # all args should have same type
                    return int(result)
                else:
                    return float(result)

        else:

            def wrapper(*args: T) -> T:
                result = func(*args)
                return type(args[0])(result)

        return wrapper

    return decorator
```

**Comment by dpaiton** [comment]
We could also potentially implement a pytorch wrapper that converts to fixed point and maintains autograd functionality. Like the other solution, this also only uses fixed-point within the function and uses float elsewhere.
```
import torch


class FixedPointFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, func, decimal_places, return_float, *args):
        fixed_args = []
        for arg in args:
            if isinstance(arg, (float, int)):
                arg = torch.tensor(arg)
            fixed_arg = to_fixed_point(arg, decimal_places)
            fixed_args.append(fixed_arg)
        result_fixed = func(*fixed_args)
        ctx.save_for_backward(result_fixed)
        ctx.decimal_places = decimal_places
        if return_float:
            result = to_floating_point(result_fixed, decimal_places)
            return result
        return result_fixed

    @staticmethod
    def backward(ctx, grad_output):
        (result_fixed,) = ctx.saved_tensors
        decimal_places = ctx.decimal_places
        grad_fixed = to_fixed_point(grad_output, decimal_places)
        grad_inputs = []
        for arg in ctx.saved_tensors:
            if arg is result_fixed:
                grad_inputs.append(grad_fixed)
            else:
                grad_inputs.append(None)
        return (None, None, None, *grad_inputs)


def fixed_point(decimal_places=18, return_float=True):
    def decorator(func):
        def wrapper(*args, **kwargs):
            torch_func = FixedPointFunction.apply
            return torch_func(func, decimal_places, return_float, *args, **kwargs)

        return wrapper

    return decorator
```

**Comment by dpaiton** [comment]
I tried to write a function that improves precision of fixedpoint rational approximations to any arbitrary level. It doesn't work, but for future solutions I'm writing it here:
```
import numpy as np

def f_fixed_point(x):
    return int(np.exp(x / 1e18) * 1e18)

def fixed_array_to_float(array):
    return np.array([val / 1e18 for val in array], dtype=float)

def get_chebyshev_coefficients(f, deg_p, deg_q, x_min, x_max):
    # Get Chebyshev coefficients for rational approximation
    x = np.linspace(x_min + 1, x_max - 1, num=1000, dtype=np.int64)
    y = np.array([f(xi) for xi in x])
    x, y = fixed_array_to_float(x), fixed_array_to_float(y)
    # Fit the numerator (P) and denominator (Q) Chebyshev polynomials
    P_coeffs = np.polynomial.chebyshev.chebfit(x, y, deg_p)
    Q_coeffs = np.polynomial.chebyshev.chebfit(x, 1 / y, deg_q)
    # Convert coefficients to fixed-point integers
    P_coeffs_fixed = [(int(c * 1e18)) for c in P_coeffs]
    Q_coeffs_fixed = [(int(c * 1e18)) for c in Q_coeffs]
    return P_coeffs_fixed, Q_coeffs_fixed
 
deg_p = 6
deg_q = 7
x_min = -42139678854452767622  # floor(log(0.5e-18)*1e18)
x_max = 135305999368893231589  # floor(log((2**255 -1) / 1e18) * 1e18)
P_coeffs, Q_coeffs = get_chebyshev_coefficients(f_fixed_point, deg_p, deg_q, x_min, x_max)

def exp_rational_approximation(x, P_coeffs, Q_coeffs):
    p = P_coeffs[-1]
    q = Q_coeffs[-1]
    for i in range(len(P_coeffs) - 2, -1, -1):
        p = (p * x) >> 96
        p += P_coeffs[i]
    for i in range(len(Q_coeffs) - 2, -1, -1):
        q = (q * x) >> 96
        q += Q_coeffs[i]
    r, _ = divmod(p, q)
    return r
```

**Comment by dpaiton** [comment]
This problem is partially addressed by #312 , which introduces the FixedPoint data type & adds the precursors necessary to convert functions to using FixedPoint arithmetic.

**Comment by dpaiton** [comment]
Fixed point refactor is well underway! Here are some issues we've come up with that will need to be addressed before we can consider the issue resolved:

- [ ] `test_calc_x_given_y`: The high-precision tests are not passing with as high of precision as we'd like
- [ ] `test_calc_out_given_in`: The regular tests are not passing with places > 8 -- @sentilesdal believes this is due to incorrect expected results.
- [ ] FixedPoint type should include the necessary dunders to make it equivalent to a `numbers.Real` type. This will allow us to use e.g. `assertAlmostEqual` with FixedPoint arguments.

**Comment by dpaiton** [comment]
Closing this issue; the tasks are largely resolved & remaining ones have their own issues.
